### ISSUE TITLE
Keep Hitting Request Limit

### ISSUE THREAD

**Issue by james090500** [McAPI/McAPI#20]
I'm trying to use the API for my website which will be receiving a shit ton of requests. I see in your API you said you have a way of not hitting the limit. Yet, every now and again I keep "Hitting the limit". Here's the error I receive.

Warning: file_get_contents(https://mcapi.de/api/user/james090500): failed to open stream: HTTP request failed! HTTP/1.1 429 Mojang Limit Reached

I have the refresh my site, and re-login a few times for it to start working again.

**Comment by jan-krueger** [comment]
Sorry, I probably have it phrased badly but I didn't intend to say that it is _impossible_ to hit the rate limit. - The API is serving around  100.000.000+ requests each month and most of them (99.5%) are hitting the _/user/-endpoint_. The Mojang API allows a max of. 4.320.000/ per IP per month.  I am _exceeding_ this limited by a __factor of 23__(!). - So I try to serve as much information as possible from the cache. The cache currently contains 682.174 unique users. 

This issue was also discussed in #18 and #19 & I am aware that this is an important issue and the system to improve the situation is in place but I am currently limited. - I am waiting for two options:

- Mojang allows access to their API via IPv6 (it is 2017, Mojang...)
- Digitalocean allows me to bind multiple IPv4 addresses to my server

If one of these things happen the issue would be resolved within hours! - I am sorry for the inconvenience. I also recommend this:

>  __What can I do to help McAPI to avoid to hit the limit?__
> Optimize your code to remove any unnecessary API calls. - If you want to optimize your code, you can cache requests in your application or store them in your own database, serialized files etc.
> Limit the request rate. Fair use, is the key word! - Yes, we love to support your application with the API. Yes, we love huge and ambitious projects and all of them are always welcome, but we also run this service for free and we want as many developers as possible to have a great time using McAPI. If you frequently notice that McAPI runs into 429 errors, you could try to distribute your requests more evenly over time.
> You can monitor this by checking how often you run into 429 error codes.
> This is a really new feature and the system is currently in beta due to the lack of testing yet. We've tested it in the past but only to a limited extent. If you run into problems feel free to create a new issue on GitHub to notify us. 

Cheers.

**Comment by james090500** [comment]
Sorry to re-open a old issue.

But its not at the point where I can't call a single request. I can on my local machine, but the server is returning the same 429 error. I am only calling 1 request and the information is being cached for around a day. So I am not making many requests at all.

**Comment by jan-krueger** [comment]
I am so sorry for the inconvenience but even in just the last hour, I had 100k requests. I think about adding additional servers, a load balancer and moving the database to dedicated servers as well, so I could load balance the application more easily but this would add a huge amount to my monthly bill & I am a student, so money is always an issue.

**Comment by james090500** [comment]
Don't worry I am a student to. I understand completely, its a free service.

It's just strange that I keep get a 429 when using php file_get_contents. But when I do a wget on the server or visit it on my website, i never get a 429 error. any ideas?

**Comment by jan-krueger** [comment]
That is actually weird... you may wanna try doing it with [curl](http://php.net/manual/de/book.curl.php) instead of file_get_contents. Here is an example:

```php
$c = curl_init();
curl_setopt_array($c, [
	CURLOPT_URL 		=> 'https://mcapi.de/api/user/Notch',
	CURLOPT_RETURNTRANSFER 	=> true
]);

$result = curl_exec($c);
curl_close($c);

if($result === false) {
	//--- Error
} else {
	echo '<pre>', var_dump(json_decode($result, true)) . '</pre>';
}
```

**Comment by james090500** [comment]
Dam, still getting a 429. And now getting it on my main PC. I am guessing you're under a lot of traffic today. Have you thought about routing through cloudflare to check traffic is legitimate?

**Comment by jan-krueger** [comment]
Yes & I had it for a year or so but it started screwing up when I implemented SSL for the website and caused endless redirections. I tried setting it up again now under http://test.mcapi.de but that gets a 502 from my gateway (PHP7.0-FPM) for some reason but the traffic not routed through Cloudflare causes no errors at all.

I got rid of the endless redirecting by following [Cloudflare's own tutorial](https://support.cloudflare.com/hc/en-us/articles/115000219871-Why-does-Flexible-SSL-cause-a-redirect-loop-) on the topic if I find a way (or maybe you have an idea) I could switch fully back to Cloudflare.

**Comment by james090500** [comment]
Are you using Nginx or Apache?

Also I am guessing you have a .htaccess somewhere telling traffic to be auto redirected to https? If so remove that. Best option is the force https on cloudflares side.

**Comment by jan-krueger** [comment]
I am using nginx. - I did disable the redirection stuff in my confs and enabled forced https on Clouflare but this caused my gateway to return a 502... so yep, it isn't working either way for some reason. I may ask in Cloudflare's forums for help.

**Comment by james090500** [comment]
Have you tried looking at nginx error logs? Also why use flexible SSL? As
that usually leaves an open connecting between cloudflare and your server.
Use full ssl.

**Comment by jan-krueger** [comment]
That solved the issue, thank you. I should have [read what each of the options means](https://support.cloudflare.com/hc/en-us/articles/200170416-What-do-the-SSL-options-mean-). I thought they were just using some fancy marketing names, well, regardless. Thanks for your help.

**Comment by james090500** [comment]
Lol no problem, also as a precaution. Disable browser integrity checks for
API as it will totally break it.

**Comment by jan-krueger** [comment]
Hi,

I just wanted to follow up with you with what the results are so far. Well, security level was set to "Low" for the first 24h in low and for the remaining days set to "Medium. CloudFlare basically flagged everything as legitimate traffic (spiked caused by CloudFlare taking some minutes to disabled "Browser Integrity Checks"). 

https://i.imgur.com/Ol6A0Vg.png (Results for the last week)

I might even increase it to "High", but I will test that if I have time to monitor traffic for a couple of hours this weekend.

Cheers & thanks for the help.

**Comment by james090500** [comment]
Hello,

That looks awesome! Hopefully should help a lot of issues. If enabled High for my security level on my entire site except the API which actually supplies users with capes (I own www.MinecraftCapes.co.uk) btw and this is my traffic report. 

https://gyazo.com/6fdd8ca55d6a8b994d3c5a7eccf0bd1f (Last 24 hours)

Cloudflare is so useful.

Because setting traffic management on the api as if a server is making lots of requests. it can be detected as spam and then the IP wont be able to be unblocked unless its solved by a captcha. 

But yet again no servers should be "exhibited threatening behavior within the last 14 days" so you should be good.
